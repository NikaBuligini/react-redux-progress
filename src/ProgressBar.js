// @flow

import React from 'react';

const MAX_PERCENT = 85;

type Props = {
  percent: number,
  updateProgress: (progress: number) => void,
  autoIncrement: boolean,
  intervalTime: number,
  color: string,
  styles?: Object,
  className?: string,
};

type State = {
  percent: number,
};

const styles = {
  wrapper: {
    position: 'fixed',
    top: '0',
    left: '0',
    width: '100%',
    transition: 'all 500mx ease-in-out',
  },
  hiddenWrapper: {
    visibility: 'hidden',
    opacity: '0',
    zIndex: '-10',
  },
  visibleWrapper: {
    visibility: 'visible',
    opacity: '1',
    zIndex: '9999',
  },
  percent: {
    transition: 'all 400ms ease',
    height: '2px',
    boxShadow: '0 0 10px rgba(119, 182, 255, 0.7)',
  },
};

function getWrapperStyles(isHidden: boolean) {
  const visibilityStyles = isHidden
    ? styles.hiddenWrapper
    : styles.visibleWrapper;

  return { ...styles.wrapper, ...visibilityStyles };
}

function getPercentStyles(
  color: string,
  percent: number,
  clientStyles: Object = {},
) {
  const customStyles = {
    width: percent <= 0 ? '0' : `${percent}%`,
    opacity: percent >= 99.9 ? '0' : '1',
    background: color,
  };

  return { ...styles.percent, ...clientStyles, ...customStyles };
}

class ProgressBar extends React.Component<Props, State> {
  static defaultProps = {
    percent: -1,
    autoIncrement: true,
    intervalTime: 450,
    color: '#77b6ff',
  };

  state = {
    percent: this.props.percent,
  };

  componentDidMount() {
    this.handleProps(this.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.interval) {
      // stop progress when new props come in.
      clearInterval(this.interval);
      this.interval = undefined;
    }

    if (this.timeout) {
      // clear timeout when new props come in.
      clearTimeout(this.timeout);
      this.timeout = undefined;
    }

    // start progress with updated props.
    this.handleProps(nextProps);
  }

  componentWillUnmount() {
    // cleaning up interval and timeout.
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = undefined;
    }

    if (this.timeout) {
      clearTimeout(this.timeout);
      this.timeout = undefined;
    }
  }

  increment = () => {
    let { percent } = this.state;

    percent += (Math.random() + 1 - Math.random()) * 3; // eslint-disable-line
    percent = percent < MAX_PERCENT ? percent : MAX_PERCENT;

    this.setState({ percent });
  };

  interval: ?number;

  handleProps = (props: Props) => {
    const { autoIncrement, percent, intervalTime } = props;

    if (autoIncrement && percent >= 0 && percent < 99) {
      this.interval = setInterval(this.increment, intervalTime);
    }

    if (percent >= 100) {
      this.setState({ percent: 99.9 }, () => {
        this.timeout = setTimeout(() => {
          this.setState({ percent: -1 }, () => this.props.updateProgress(-1));
        }, 400);
      });
    } else {
      this.setState({ percent });
    }
  };

  timeout: ?number;

  render() {
    const { percent } = this.state;

    // Hide progress bar if percent is less than 0.
    const isHidden = percent < 0 || percent >= 100;

    // Set `state.percent` as width.
    return (
      <div style={getWrapperStyles(isHidden)}>
        <div
          className={this.props.className}
          style={getPercentStyles(this.props.color, percent, this.props.styles)}
        />
      </div>
    );
  }
}

export default ProgressBar;
