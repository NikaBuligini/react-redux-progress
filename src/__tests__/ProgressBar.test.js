import React from 'react';
import sinon from 'sinon';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import ProgressBar from '../ProgressBar';

let clock = null;

describe('<ProgressBar />', () => {
  beforeEach(() => {
    clock = sinon.useFakeTimers();
  });

  afterEach(() => {
    clock.restore();
  });

  it('should initially render progress bar', () => {
    const component = renderer.create(<ProgressBar />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with color name', () => {
    const component = renderer.create(<ProgressBar color="red" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with color hex', () => {
    const component = renderer.create(<ProgressBar color="#ff0000" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with styles', () => {
    const component = renderer.create(
      <ProgressBar styles={{ height: '5px' }} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with className', () => {
    const component = renderer.create(
      <ProgressBar className="my-custom-classname" />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar using passed color', () => {
    const component = renderer.create(<ProgressBar color="#db7093" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should set state.percent as props.percent', () => {
    const expected = 50;
    const renderedComponent = mount(<ProgressBar percent={expected} />);
    expect(renderedComponent.state().percent).toEqual(expected);
  });

  it('should call componentDidMount', () => {
    sinon.spy(ProgressBar.prototype, 'componentDidMount');
    // eslint-disable-next-line
    const renderedComponent = mount(
      <ProgressBar percent={0} updateProgress={noop => noop} />,
    );
    expect(ProgressBar.prototype.componentDidMount.calledOnce).toEqual(true);
    ProgressBar.prototype.componentDidMount.restore();
  });

  it('should call componentWillReceiveProps', () => {
    sinon.spy(ProgressBar.prototype, 'componentWillReceiveProps');
    const renderedComponent = mount(
      <ProgressBar percent={0} updateProgress={noop => noop} />,
    );
    renderedComponent.setProps({ percent: 50 });
    expect(ProgressBar.prototype.componentWillReceiveProps.calledOnce).toEqual(
      true,
    );
    ProgressBar.prototype.componentWillReceiveProps.restore();
  });

  it('should unset ProgressBar.interval after getting new props', () => {
    const renderedComponent = mount(
      <ProgressBar percent={0} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.interval).toBeDefined();
    inst.componentWillReceiveProps({ percent: 50 });
    expect(inst.interval).toBeUndefined();
  });

  it('should unset ProgressBar.timeout after getting new props', () => {
    const renderedComponent = mount(
      <ProgressBar percent={100} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.timeout).toBeDefined();
    inst.componentWillReceiveProps({ percent: 50 });
    expect(inst.timeout).toBeUndefined();
  });

  it('should set state to -1 after new route mounts', () => {
    const renderedComponent = mount(
      <ProgressBar percent={0} updateProgress={noop => noop} />,
    );
    renderedComponent.setProps({ percent: 100 });
    clock.tick(501);
    expect(renderedComponent.state().percent).toEqual(-1);
  });

  it('should call componentWillUnmount', () => {
    sinon.spy(ProgressBar.prototype, 'componentWillUnmount');
    const renderedComponent = mount(
      <ProgressBar percent={0} updateProgress={noop => noop} />,
    );
    renderedComponent.unmount();
    expect(ProgressBar.prototype.componentWillUnmount.calledOnce).toEqual(true);
    ProgressBar.prototype.componentWillUnmount.restore();
  });

  it('should unset ProgressBar.interval after unmounting', () => {
    sinon.spy(ProgressBar.prototype, 'componentWillUnmount');
    const renderedComponent = mount(
      <ProgressBar percent={0} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.interval).toBeDefined();
    renderedComponent.unmount();
    expect(inst.interval).toBeUndefined();
    ProgressBar.prototype.componentWillUnmount.restore();
  });

  it('should unset ProgressBar.timeout after unmounting', () => {
    sinon.spy(ProgressBar.prototype, 'componentWillUnmount');
    const renderedComponent = mount(
      <ProgressBar percent={100} updateProgress={noop => noop} />,
    );
    const inst = renderedComponent.instance();

    clock.tick(1000);
    expect(inst.timeout).toBeDefined();
    renderedComponent.unmount();
    expect(inst.timeout).toBeUndefined();
    ProgressBar.prototype.componentWillUnmount.restore();
  });

  describe('increment progress', () => {
    beforeEach(() => {
      clock = sinon.useFakeTimers();
    });

    afterEach(() => {
      clock.restore();
    });

    it('should start incrementing progress if 0 <= percent < 100', () => {
      const initialPercent = 50;
      const renderedComponent = mount(
        <ProgressBar percent={initialPercent} updateProgress={noop => noop} />,
      );
      clock.tick(1000);
      expect(renderedComponent.state().percent).toBeGreaterThan(initialPercent);
    });

    it('should stop incrementing progress if percent >= 100', () => {
      const initialPercent = 100;
      const expected = -1;
      const renderedComponent = mount(
        <ProgressBar percent={initialPercent} updateProgress={noop => noop} />,
      );
      clock.tick(1000);
      expect(renderedComponent.state().percent).toEqual(expected);
    });
  });
});
