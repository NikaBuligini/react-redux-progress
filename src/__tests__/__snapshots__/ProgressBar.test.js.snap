// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> should initially render progress bar 1`] = `
<div
  style={
    Object {
      "left": "0",
      "opacity": "0",
      "position": "fixed",
      "top": "0",
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": "-10",
    }
  }
>
  <div
    className={undefined}
    style={
      Object {
        "background": "#77b6ff",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0",
      }
    }
  />
</div>
`;

exports[`<ProgressBar /> should render progress bar using passed color 1`] = `
<div
  style={
    Object {
      "left": "0",
      "opacity": "0",
      "position": "fixed",
      "top": "0",
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": "-10",
    }
  }
>
  <div
    className={undefined}
    style={
      Object {
        "background": "#db7093",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0",
      }
    }
  />
</div>
`;

exports[`<ProgressBar /> should render progress bar with className 1`] = `
<div
  style={
    Object {
      "left": "0",
      "opacity": "0",
      "position": "fixed",
      "top": "0",
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": "-10",
    }
  }
>
  <div
    className="my-custom-classname"
    style={
      Object {
        "background": "#77b6ff",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0",
      }
    }
  />
</div>
`;

exports[`<ProgressBar /> should render progress bar with color 1`] = `
<div
  style={
    Object {
      "left": "0",
      "opacity": "0",
      "position": "fixed",
      "top": "0",
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": "-10",
    }
  }
>
  <div
    className={undefined}
    style={
      Object {
        "background": "red",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0",
      }
    }
  />
</div>
`;

exports[`<ProgressBar /> should render progress bar with styles 1`] = `
<div
  style={
    Object {
      "left": "0",
      "opacity": "0",
      "position": "fixed",
      "top": "0",
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": "-10",
    }
  }
>
  <div
    className={undefined}
    style={
      Object {
        "background": "#77b6ff",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "5px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0",
      }
    }
  />
</div>
`;
